// @signalwire/webrtc/dist/cjs/core/src/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.INTERNAL_GLOBAL_VIDEO_EVENTS = exports.PRODUCT_PREFIXES = exports.GLOBAL_VIDEO_EVENTS = exports.PRODUCT_PREFIX_VOICE_CALL = exports.PRODUCT_PREFIX_VOICE = exports.PRODUCT_PREFIX_MESSAGING = exports.PRODUCT_PREFIX_TASK = exports.PRODUCT_PREFIX_PUBSUB = exports.PRODUCT_PREFIX_CHAT = exports.PRODUCT_PREFIX_VIDEO_MANAGER = exports.PRODUCT_PREFIX_VIDEO = exports.SYNTHETIC_EVENT_PREFIX = exports.LOCAL_EVENT_PREFIX = exports.EVENT_NAMESPACE_DIVIDER = exports.WebSocketState = exports.DEFAULT_HOST = exports.SESSION_ID = exports.REMOVE = exports.ADD = exports.STORAGE_PREFIX = void 0;
exports.STORAGE_PREFIX = '@signalwire:';
exports.ADD = 'add';
exports.REMOVE = 'remove';
exports.SESSION_ID = 'sessId';
exports.DEFAULT_HOST = 'wss://relay.signalwire.com';
var WebSocketState;
(function (WebSocketState) {
    WebSocketState[WebSocketState["CONNECTING"] = 0] = "CONNECTING";
    WebSocketState[WebSocketState["OPEN"] = 1] = "OPEN";
    WebSocketState[WebSocketState["CLOSING"] = 2] = "CLOSING";
    WebSocketState[WebSocketState["CLOSED"] = 3] = "CLOSED";
})(WebSocketState = exports.WebSocketState || (exports.WebSocketState = {}));
/**
 * Used for namespacing events.
 */
exports.EVENT_NAMESPACE_DIVIDER = ':';
exports.LOCAL_EVENT_PREFIX = '__local__';
exports.SYNTHETIC_EVENT_PREFIX = '__synthetic__';
exports.PRODUCT_PREFIX_VIDEO = 'video';
/**
 * video-manager is an [internal] superset of the video apis
 */
exports.PRODUCT_PREFIX_VIDEO_MANAGER = 'video-manager';
exports.PRODUCT_PREFIX_CHAT = 'chat';
/**
 * For now both, `PubSub` and `Chat` share the same
 * namespace but this might change in the future.
 */
exports.PRODUCT_PREFIX_PUBSUB = 'chat';
exports.PRODUCT_PREFIX_TASK = 'tasking';
exports.PRODUCT_PREFIX_MESSAGING = 'messaging';
exports.PRODUCT_PREFIX_VOICE = 'voice';
exports.PRODUCT_PREFIX_VOICE_CALL = 'calling';
exports.GLOBAL_VIDEO_EVENTS = ['room.started', 'room.ended'];
exports.PRODUCT_PREFIXES = [
    exports.PRODUCT_PREFIX_VIDEO,
    exports.PRODUCT_PREFIX_VIDEO_MANAGER,
    exports.PRODUCT_PREFIX_CHAT,
    exports.PRODUCT_PREFIX_PUBSUB,
    exports.PRODUCT_PREFIX_TASK,
    exports.PRODUCT_PREFIX_MESSAGING,
    exports.PRODUCT_PREFIX_VOICE,
    exports.PRODUCT_PREFIX_VOICE_CALL,
];
/**
 * For internal usage only. These are the fully qualified event names
 * sent by the server
 * @internal
 */
exports.INTERNAL_GLOBAL_VIDEO_EVENTS = exports.GLOBAL_VIDEO_EVENTS.map((event) => `${exports.PRODUCT_PREFIX_VIDEO}.${event}`);
