// @signalwire/webrtc/dist/cjs/core/src/utils/extendComponent.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const extendComponent_1 = require("./extendComponent");
describe('extendComponent', () => {
    it('should allow us to extend an classes with new methods', () => {
        class A {
        }
        const ExtendedClass = (0, extendComponent_1.extendComponent)(A, {
            methodA: {},
            methodB: {},
            methodC: {},
            methodD: {},
        });
        const instance = new ExtendedClass();
        expect(instance).toHaveProperty('methodA');
        expect(instance).toHaveProperty('methodB');
        expect(instance).toHaveProperty('methodC');
        expect(instance).toHaveProperty('methodD');
    });
    it('should throw if the base class already has the method defined', () => {
        expect(() => (0, extendComponent_1.extendComponent)(class {
            methodA() { }
        }, {
            methodA: {},
            methodB: {},
            methodC: {},
            methodD: {},
        })).toThrow();
    });
});
