// @signalwire/webrtc/dist/mjs/core/src/rooms/RoomSessionRecording.js
import { connect } from '../redux';
import { BaseComponent } from '../BaseComponent';
export class RoomSessionRecordingAPI extends BaseComponent {
    constructor(options) {
        super(options);
        this._payload = options.payload;
    }
    get id() {
        return this._payload.recording.id;
    }
    get roomId() {
        return this._payload.room_id;
    }
    get roomSessionId() {
        return this._payload.room_session_id;
    }
    get state() {
        return this._payload.recording.state;
    }
    get duration() {
        return this._payload.recording.duration;
    }
    get startedAt() {
        if (!this._payload.recording.started_at)
            return undefined;
        return new Date(this._payload.recording.started_at * 1000);
    }
    get endedAt() {
        if (!this._payload.recording.ended_at)
            return undefined;
        return new Date(this._payload.recording.ended_at * 1000);
    }
    /** @internal */
    setPayload(payload) {
        this._payload = payload;
    }
    async pause() {
        await this.execute({
            method: 'video.recording.pause',
            params: {
                room_session_id: this.getStateProperty('roomSessionId'),
                recording_id: this.getStateProperty('id'),
            },
        });
    }
    async resume() {
        await this.execute({
            method: 'video.recording.resume',
            params: {
                room_session_id: this.getStateProperty('roomSessionId'),
                recording_id: this.getStateProperty('id'),
            },
        });
    }
    async stop() {
        await this.execute({
            method: 'video.recording.stop',
            params: {
                room_session_id: this.getStateProperty('roomSessionId'),
                recording_id: this.getStateProperty('id'),
            },
        });
    }
}
export const createRoomSessionRecordingObject = (params) => {
    const recording = connect({
        store: params.store,
        Component: RoomSessionRecordingAPI,
    })(params);
    return recording;
};
