// @signalwire/webrtc/dist/cjs/core/src/types/videoMember.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MEMBER_UPDATED_EVENTS = exports.MEMBER_UPDATABLE_PROPS = exports.INTERNAL_MEMBER_UPDATED_EVENTS = exports.INTERNAL_MEMBER_UPDATABLE_PROPS = void 0;
const constants_1 = require("../utils/constants");
const utils_1 = require("../utils");
/**
 * Used to not duplicate member fields across constants and types
 * and generate `MEMBER_UPDATED_EVENTS` below.
 * `key`: `type`
 */
exports.INTERNAL_MEMBER_UPDATABLE_PROPS = {
    audio_muted: true,
    video_muted: true,
    deaf: true,
    visible: true,
    input_volume: 1,
    output_volume: 1,
    input_sensitivity: 1,
};
exports.INTERNAL_MEMBER_UPDATED_EVENTS = Object.keys(exports.INTERNAL_MEMBER_UPDATABLE_PROPS).map((key) => {
    return `${constants_1.PRODUCT_PREFIX_VIDEO}.member.updated.${key}`;
});
exports.MEMBER_UPDATABLE_PROPS = (0, utils_1.toExternalJSON)(exports.INTERNAL_MEMBER_UPDATABLE_PROPS);
exports.MEMBER_UPDATED_EVENTS = Object.keys(exports.MEMBER_UPDATABLE_PROPS).map((key) => {
    return `member.updated.${key}`;
});
