// @signalwire/webrtc/dist/cjs/core/src/redux/features/component/componentSlice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.componentReducer = exports.componentActions = exports.initialComponentState = void 0;
const createDestroyableSlice_1 = require("../../utils/createDestroyableSlice");
exports.initialComponentState = {
    byId: {},
};
const componentSlice = (0, createDestroyableSlice_1.createDestroyableSlice)({
    name: 'components',
    initialState: exports.initialComponentState,
    reducers: {
        upsert: (state, { payload }) => {
            if (payload.id in state.byId) {
                return Object.assign(Object.assign({}, state), { byId: Object.assign(Object.assign({}, state.byId), { [payload.id]: Object.assign(Object.assign({}, state.byId[payload.id]), payload) }) });
            }
            else {
                return Object.assign(Object.assign({}, state), { byId: Object.assign(Object.assign({}, state.byId), { [payload.id]: payload }) });
            }
        },
        cleanup: (state, { payload }) => {
            return Object.assign(Object.assign({}, state), { byId: Object.entries(state.byId).reduce((reducer, [componentId, value]) => {
                    if (!payload.ids.includes(componentId)) {
                        reducer[componentId] = value;
                    }
                    return reducer;
                }, {}) });
        },
    },
});
// prettier-ignore
exports.componentActions = componentSlice.actions, exports.componentReducer = componentSlice.reducer;
