// @signalwire/webrtc/dist/cjs/core/src/redux/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAction = exports.getCustomSagaActionType = exports.makeCustomSagaAction = exports.sessionForceCloseAction = exports.sessionReconnectingAction = exports.sessionDisconnectedAction = exports.socketMessageAction = exports.authExpiringAction = exports.authSuccessAction = exports.authErrorAction = exports.reauthAction = exports.destroyAction = exports.initAction = void 0;
const toolkit_1 = require("./toolkit");
Object.defineProperty(exports, "createAction", { enumerable: true, get: function () { return toolkit_1.createAction; } });
exports.initAction = (0, toolkit_1.createAction)('swSdk/init');
exports.destroyAction = (0, toolkit_1.createAction)('swSdk/destroy');
/**
 * Used to trigger a `signalwire.reauthenticate`
 */
exports.reauthAction = (0, toolkit_1.createAction)('swSdk/reauth');
exports.authErrorAction = (0, toolkit_1.createAction)('auth/error');
exports.authSuccessAction = (0, toolkit_1.createAction)('auth/success');
exports.authExpiringAction = (0, toolkit_1.createAction)('auth/expiring');
exports.socketMessageAction = (0, toolkit_1.createAction)('socket/message');
exports.sessionDisconnectedAction = (0, toolkit_1.createAction)('session.disconnected');
exports.sessionReconnectingAction = (0, toolkit_1.createAction)('session.reconnecting');
exports.sessionForceCloseAction = (0, toolkit_1.createAction)('session.forceClose');
const formatCustomSagaAction = (id, action) => {
    return `${action.type}/${id}`;
};
const makeCustomSagaAction = (id, action) => {
    return Object.assign(Object.assign({}, action), { type: formatCustomSagaAction(id, action) });
};
exports.makeCustomSagaAction = makeCustomSagaAction;
const getCustomSagaActionType = (id, action) => {
    return formatCustomSagaAction(id, action);
};
exports.getCustomSagaActionType = getCustomSagaActionType;
